### PR TITLE
test: Spectron (Initial commit)

### PR DESCRIPTION
**Pull request by KiranNiranjan**
### Spectron

- Currently, sandbox is **disabled** if running in `NODE_ENV`
- Wrote few testes cases (Yet to verify on Windows OS)
- `appveyor` WIP (verifying if this workout)

### Test added

- About app
- Fullscreen
- Minimize
- Notification
- Screen picker
- Screen sharing indicator
- Spellchecker

Screenshot 2019-12-23 at 3 08 02 PM

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+os: unstable
+
+platform:
+  - x86
+  - x64
+
+branches:
+  only:
+    - master
```
Comment (VishwasShashidhar) [inline appveyor.yml]: I think we should run it against all branches & PRs.

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+skip_tags: true
```
Comment (VishwasShashidhar) [inline appveyor.yml]: we should also run it against tags as well since we'll be using them for releases

File: appveyor.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  - node_modules -> package.json
+
+install:
+  - ps: Install-Product node $env:nodejs_version
+  - npm install
+
+test_script:
+  - node --version
+  - npm --version
+  - npm run compile:spec && npm run test:spectron
```
Comment (VishwasShashidhar) [inline appveyor.yml]: Let's also run unit tests here before running the automation tests.

File: spectron/about-app.spec.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    await stopApplication(app);
+});
+
+test('about-app: verify application minimize / maximize feature', async (t) => {
+    robotActions.clickAppMenu();
+    robot.keyTap('down');
+    robot.keyTap('enter');
+
+    // wait for about window to load
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/about-app.spec.ts]: should we add this to a variable?

File: spectron/about-app.spec.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    await app.client.click('.AboutApp-copy-button');
+    const clipboard = JSON.parse(await app.client.electron.remote.clipboard.readText());
+
+    t.log(clipboard);
+    t.true(clipboard.hasOwnProperty('appName'));
+    t.true(clipboard.hasOwnProperty('clientVersion'));
+    t.true(clipboard.hasOwnProperty('sfeVersion'));
+    t.true(clipboard.hasOwnProperty('sdaVersion'));
+    t.true(clipboard.hasOwnProperty('sdaBuildNumber'));
+    robotActions.closeWindow();
```
Comment (VishwasShashidhar) [inline spectron/about-app.spec.ts]: is there a way to compare the entire json perhaps?

File: spectron/fixtures/robot-actions.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import * as robot from 'robotjs';
+import { isMac } from '../../src/common/env';
+
+class RobotActions {
+
+    constructor() {
+        robot.setKeyboardDelay(1000);
```
Comment (VishwasShashidhar) [inline spectron/fixtures/robot-actions.ts]: should we move this to a variable?

File: spectron/fixtures/robot-actions.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        robot.setMouseDelay(1000);
```
Comment (VishwasShashidhar) [inline spectron/fixtures/robot-actions.ts]: should we move this to a variable?

File: spectron/full-screen.spec.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    t.true(app.isRunning());
+});
+
+test.after.always(async () => {
+    await stopApplication(app);
+});
+
+test('fullscreen: verify application full screen feature', async (t) => {
+    await app.browserWindow.loadURL(getDemoFilePath());
+    await app.client.waitUntilWindowLoaded(5000);
```
Comment (VishwasShashidhar) [inline spectron/full-screen.spec.ts]: should we move this to a variable?

File: spectron/full-screen.spec.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    robotActions.toggleFullscreen();
+    t.true(await app.browserWindow.isFullScreen());
+
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/full-screen.spec.ts]: should we move this to a variable?

File: spectron/notification.spec.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    await stopApplication(app);
+});
+
+test('Verify is the application is running', async (t) => {
+    t.true(app.isRunning());
+});
+
+test('Verify notification window is created', async (t) => {
+    await app.browserWindow.loadURL(getDemoFilePath());
+    await app.client.waitUntilWindowLoaded(5000);
```
Comment (VishwasShashidhar) [inline spectron/notification.spec.ts]: should we move this to a variable?

File: spectron/notification.spec.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    await app.client.click('#notf');
+
+    await sleep(1000);
```
Comment (VishwasShashidhar) [inline spectron/notification.spec.ts]: should we move this to a variable?

File: spectron/notification.spec.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    t.timeout(10000);
```
Comment (VishwasShashidhar) [inline spectron/notification.spec.ts]: should we move this to a variable?

File: spectron/notification.spec.ts [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    t.is(await app.client.getWindowCount(), 2);
+    await app.client.windowByIndex(1);
+    await app.client.click('.close');
+
+    await sleep(2000);
```
Comment (VishwasShashidhar) [inline spectron/notification.spec.ts]: should we move this to a variable?

File: spectron/notification.spec.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    await app.client.windowByIndex(0);
+});
+
+test('Verify notification window is hidden', async (t) => {
+    await app.client.click('#notf');
+
+    await sleep(1000);
+    t.timeout(5000);
```
Comment (VishwasShashidhar) [inline spectron/notification.spec.ts]: should we move this to a variable?

File: spectron/notification.spec.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    await app.client.windowByIndex(1);
+    await app.client.click('.close');
+
+    await sleep(2000);
```
Comment (VishwasShashidhar) [inline spectron/notification.spec.ts]: should we move this to a variable?

File: spectron/screen-picker.spec.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+import { getDemoFilePath, sleep, startApplication, stopApplication } from './fixtures/spectron-setup';
+
+let app;
+
+export const openScreenPicker = async (window) => {
+    if (!window) {
+        throw new Error('openScreenPicker: must be called with Application');
+    }
+    await window.client.scroll(125, 1000);
```
Comment (VishwasShashidhar) [inline spectron/screen-picker.spec.ts]: should we move this to a variable?

File: spectron/screen-picker.spec.ts [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    await sleep(500);
+    await window.client.click('#get-sources');
+    await window.client.waitUntilWindowLoaded(5000);
```
Comment (VishwasShashidhar) [inline spectron/screen-picker.spec.ts]: should we move this to a variable?

File: spectron/screen-picker.spec.ts [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+import { getDemoFilePath, sleep, startApplication, stopApplication } from './fixtures/spectron-setup';
+
+let app;
+
+export const openScreenPicker = async (window) => {
+    if (!window) {
+        throw new Error('openScreenPicker: must be called with Application');
+    }
+    await window.client.scroll(125, 1000);
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/screen-picker.spec.ts]: should we move this to a variable?

File: spectron/screen-picker.spec.ts [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+test.after.always(async () => {
+    await stopApplication(app);
+});
+
+test('screen-picker: verify screen-picker close button', async (t) => {
+    await app.browserWindow.loadURL(getDemoFilePath());
+    await app.client.waitUntilWindowLoaded(5000);
+    await openScreenPicker(app);
+
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/screen-picker.spec.ts]: should we move this to a variable?

File: spectron/screen-picker.spec.ts [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    t.is(await app.client.getWindowCount(), 2);
+    await app.client.windowByIndex(1);
+    await app.client.click('.ScreenPicker-x-button');
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/screen-picker.spec.ts]: should we move this to a variable?

File: spectron/screen-picker.spec.ts [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+    t.is(await app.client.getWindowCount(), 1);
+});
+
+test('screen-picker: verify screen-picker escape keyboard actions', async (t) => {
+    await app.client.windowByIndex(0);
+    await openScreenPicker(app);
+
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/screen-picker.spec.ts]: should we move this to a variable?

File: spectron/screen-picker.spec.ts [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+    t.is(await app.client.getWindowCount(), 2);
+    robot.keyTap('escape');
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/screen-picker.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+    stopApplication,
+} from './fixtures/spectron-setup';
+
+let app;
+
+export const openScreenPicker = async (window) => {
+    if (!window) {
+        throw new Error('openScreenPicker: must be called with Application');
+    }
+    await window.client.scroll(125, 1000);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+    await window.client.click('#get-sources');
+    await window.client.waitUntilWindowLoaded(5000);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+    t.true(app.isRunning());
+});
+
+test.after.always(async () => {
+    await stopApplication(app);
+});
+
+test('screen-sharing-indicator: verify screen sharing indicator with frame is shown', async (t) => {
+    await app.browserWindow.loadURL(getDemoFilePath());
+    await app.client.waitUntilWindowLoaded(5000);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+    await openScreenPicker(app);
+    robot.setKeyboardDelay(2000);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+    t.is(await app.client.getWindowCount(), 2);
+    await app.client.windowByIndex(1);
+    // will select the entire screen option in the picker
+    robot.keyTap('right');
+    robot.keyTap('left');
+    robot.keyTap('enter');
+
+    await sleep(2000);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+    t.is(await app.client.getWindowCount(), 3);
+});
+
+test('screen-sharing-indicator: verify screen sharing indicator title', async (t) => {
+    // including the screen sharing frame
+    await app.client.windowByIndex(2);
+    await app.client.waitUntilWindowLoaded(5000);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/screen-sharing-indicator.spec.ts [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+
+    const indicatorTitle = await app.browserWindow.getTitle();
+    if (indicatorTitle !== 'Screen Sharing Indicator - Symphony') {
+        await app.client.windowByIndex(1);
+    }
+
+    await app.client.click('.stop-sharing-button');
+    await app.client.windowByIndex(0);
+    await sleep(500);
```
Comment (VishwasShashidhar) [inline spectron/screen-sharing-indicator.spec.ts]: should we move this to a variable?

File: spectron/spell-checker.spec.ts [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+    app = await startApplication() as Application;
+    t.true(app.isRunning());
+});
+
+test.after.always(async () => {
+    await stopApplication(app);
+});
+
+test('spell-checker: verify application spell checking feature', async (t) => {
+    robot.setKeyboardDelay(1000);
```
Comment (VishwasShashidhar) [inline spectron/spell-checker.spec.ts]: should we move this to a variable?

File: spectron/spell-checker.spec.ts [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+    const missSpelledWord = 'teest ';
+
+    await app.browserWindow.loadURL(getDemoFilePath());
+    await app.client.waitUntilWindowLoaded(5000);
```
Comment (VishwasShashidhar) [inline spectron/spell-checker.spec.ts]: should we move this to a variable?

File: spectron/spell-checker.spec.ts [hunk 34]
```diff
@@ -0,0 +1,37 @@
...
+    await app.client.electron.remote.clipboard.writeText(missSpelledWord);
+    await app.client.click('#tag');
+    await app.client.webContents.paste();
+    await app.client.waitForValue('#tag', 5000);
```
Comment (VishwasShashidhar) [inline spectron/spell-checker.spec.ts]: should we move this to a variable?

File: spectron/about-app.spec.ts (hunk 5 above)
Comment (KiranNiranjan) [reply spectron/about-app.spec.ts]: I don't think so, as these values keep changing.

**Comment by VishwasShashidhar** [comment]
@KiranNiranjan Is this actually relevant right now? Will you create a new PR or continue with this?

**Comment by KiranNiranjan** [comment]
@VishwasShashidhar @johankwarnmarksymphony @mattias-symphony This PR is ready. Here is a summary of the changes. Please review again. Thanks

1. Unit test runs on GitHub actions for (Ubuntu, macOS, Windows)
2. Spectron tests run on Travis for (macOS)

[WIP] Spectron - Windows (Will work on investigating and create a separate PR)
[WIP] Spectron - Linux (Will work on investigating and create a separate PR)

Screenshot 2020-03-12 at 10 08 52 PM

**Pull request merged**